vaadin-cdi/src/main/java/com/vaadin/cdi/context/RouteScopedContext.java

```diff
@@ -94,29 +268,58 @@ public boolean isActive() {
 
     @Override
     protected ContextualStorage getContextualStorage(Contextual<?> contextual,
-                                                     boolean createIfNotExist) {
-        Class key = convertToKey(contextual);
+            boolean createIfNotExist) {
+        RouteStorageKey key = convertToKey(contextual);
         return contextManager.getContextualStorage(key, createIfNotExist);
     }
 
-    private Class convertToKey(Contextual<?> contextual) {
-        if (!(contextual instanceof Bean)) {
-            if (contextual instanceof PassivationCapable) {
-                final String id = ((PassivationCapable) contextual).getId();
-                contextual = beanManager.getPassivationCapableBean(id);
-            } else {
-                throw new IllegalArgumentException(
-                        contextual.getClass().getName()
-                                + " is not of type " + Bean.class.getName());
-            }
+    private RouteStorageKey convertToKey(Contextual<?> contextual) {
+        Bean<?> bean = getBean(contextual);
+        UI ui = UI.getCurrent();
+        Class<?> owner = getOwner(ui, bean);
+        if (!navigationChainHasOwner(ui, owner)) {
+            throw new IllegalStateException(String.format(
+                    "Route owner '%s' instance is not available in the "
+                            + "active navigation components chain: the scope defined by the bean '%s' doesn't exist.",
+                    owner, bean.getBeanClass()));
         }
-        final Bean<?> bean = (Bean<?>) contextual;
-        return bean.getQualifiers()
-                .stream()
+        return contextManager.getKey(ui, owner);
+    }
+
+    private boolean navigationChainHasOwner(UI ui, Class<?> owner) {
+        NavigationData data = ComponentUtil.getData(ui, NavigationData.class);
+        if (owner.equals(data.getNavigationTarget())) {
+            return true;
+        }
+        return data.getLayouts().stream()
+                .anyMatch(clazz -> clazz.equals(owner));
+    }
+
+    @SuppressWarnings("unchecked")
+    private Class<?> getOwner(UI ui, Bean<?> bean) {
+        return bean.getQualifiers().stream()
                 .filter(annotation -> annotation instanceof RouteScopeOwner)
-                .map(annotation -> (Class) (((RouteScopeOwner) annotation).value()))
-                .findFirst()
-                .orElse(bean.getBeanClass());
+                .map(annotation -> (Class<?>) (((RouteScopeOwner) annotation)
+                        .value()))
+                .findFirst().orElseGet(() -> getCurrentNavigationTarget(ui));
     }
 
+    @SuppressWarnings("rawtypes")
+    private Class getCurrentNavigationTarget(UI ui) {
+        NavigationData data = ComponentUtil.getData(ui, NavigationData.class);
```

Comment:
Is it possible that this will be `null` ? If something tries to call `getContextualStorage` -> `convertToKey` -> `getOwner` -> `getCurrentNavigationTarget`, maybe because a route scoped bean is injected somewhere before the route has been initialized and before the `UI` has any `NavigationData` set, then it would fail here for NPE instead of in `convertToKey` that checks for ` if (!navigationChainHasOwner(ui, owner)) {`.

Reply:
done